evalml/pipelines/component_graph.py

```diff
@@ -0,0 +1,295 @@
+import networkx as nx
+import pandas as pd
+from networkx.algorithms.dag import topological_sort
+
+from evalml.model_family import ModelFamily
+from evalml.pipelines.components import ComponentBase
+from evalml.pipelines.components.utils import handle_component_class
+from evalml.utils import import_or_raise
+
+
+class ComponentGraph:
+    def __init__(self, component_dict=None, random_state=0):
+        """ Initializes a component graph for a pipeline as a DAG.
+        """
+        self.component_dict = component_dict or {}
+        for key, value in self.component_dict.items():
+            if not isinstance(value, list):
+                self.component_dict[key] = [value]
+        self._compute_order = []
+        self._recompute_order()
+        self.random_state = random_state
+
+    def from_list(self, component_list):
```

Comment:
Unfortunately, we do need to call `handle_component_class` on the key. The current `component_graph` argument can be either a list of strings or classes, but we need to ensure the keys are all strings and not classes. Along the same vein, the list of parents has to map to valid components in the graph, and since we call `handle_component_class` on those values, we need to call `handle_component_class` when naming each component to ensure they match.

Reply:
With the number of `handle_component_class` calls that are necessary to implement this with list comprehension, I'm leaning toward leaving this closer to my original implementation just for readability's sake. This also avoids the problem with the code you suggested where the final component is listed as the parent of the first component. Of course, if you think list comprehension is still a better choice just let me know, I can do that as well! 

This would look more like:
```component_dict = {handle_component_class(el).name: [handle_component_class(el), f"{handle_component_class(component_list[idx - 1]).name}.x"] for idx, el in enumerate(component_list)}```